src/PostgREST/Plan.hs

```diff
@@ -604,6 +604,22 @@ findRel schema allRels origin target hint =
             )
       ) $ fromMaybe mempty $ HM.lookup (QualifiedIdentifier schema origin, schema) allRels
 
+-- Add JsonAgg aggregates to selected fields that do not have other aggregates and:
+-- * Belong to a spread to-many relationship
+-- * Are to-one spread but are nested inside a spread to-many relationship
+addJsonAggToManySpread :: Bool -> ReadPlanTree -> Either ApiRequestError ReadPlanTree
+addJsonAggToManySpread isNestedInToManyRel (Node rp@ReadPlan{select, relIsSpread, relToParent} forest) =
```

Comment:
I don't think it's practical to have cases like this:

```http
GET /entities?select=...child_entites(grandchild_entities(name))
```

So it should return an error for these cases. Only spread relationships are allowed to be nested inside to-many spread relationships.

Reply:
> Hm. I would expect to get something like this back, no?

Now that you mention it, yes, something like that could be expected. This would have the same caveats that I mentioned here https://github.com/PostgREST/postgrest/pull/3640#issuecomment-2232120378. For example, if there was a M2M between `child_entities` and `grandchild_entities` then repeated values can be expected and a `DISTINCT` would be needed.

Something like this:

<details><summary>Old query before the implementation</summary>
<p>

```sql
SELECT "entities_child_entities_1"."grandchild_entities"
FROM "test"."entities"
LEFT JOIN LATERAL
  (SELECT COALESCE("child_entities_grandchild_entities_2"."child_entities_grandchild_entities_2", '[]') AS "grandchild_entities"
   FROM "test"."child_entities" AS "child_entities_1"
   LEFT JOIN LATERAL
     (SELECT json_agg("child_entities_grandchild_entities_2")::jsonb AS "child_entities_grandchild_entities_2"
      FROM
        (SELECT "grandchild_entities_2"."name"
         FROM "test"."grandchild_entities" AS "grandchild_entities_2"
         WHERE "grandchild_entities_2"."parent_id" = "child_entities_1"."id") AS "child_entities_grandchild_entities_2") AS "child_entities_grandchild_entities_2" ON TRUE
   WHERE "child_entities_1"."parent_id" = "test"."entities"."id") AS "entities_child_entities_1" ON TRUE;
```

</p>
</details> 

<details><summary>New query after the implementation</summary>
<p>

```sql
SELECT json_agg(distinct "entities_child_entities_1"."grandchild_entities") AS grandchild_entitites
FROM "api"."entities"
LEFT JOIN LATERAL
  (SELECT "child_entities_grandchild_entities_2" AS "grandchild_entities"
   FROM "api"."child_entities" AS "child_entities_1"
   LEFT JOIN LATERAL
     (SELECT "grandchild_entities_2"."name"
      FROM "api"."grandchild_entities" AS "grandchild_entities_2"
      WHERE "grandchild_entities_2"."parent_id" = "child_entities_1"."id") AS "child_entities_grandchild_entities_2" ON TRUE
   WHERE "child_entities_1"."parent_id" = "api"."entities"."id") AS "entities_child_entities_1" ON TRUE;
```

</p>
</details> 

I simplified the `SELECT` to show the bigger picture better; but if we don't want to include `null`s in the response, then the selected embed would look like:

```sql
SELECT
  coalesce(
    json_agg(distinct "entities_child_entities_1"."grandchild_entities") filter (where "entities_child_entities_1"."grandchild_entities" is not null)
    ,'[]'
  ) AS grandchild_entitites
```

I'm not sure if it'd be less performant if there was further embeds to `grandchild_entities`. Perhaps, there's an alternative query or approach that I might be missing.

Reply:
This is important for something like `posts -> post_tag -> tags`, and you want to simply say:
```
posts?select=*,...post_tag(tags(*))
```
In which case I would expect to get back:
```js
posts: [{
  tags: [{
    name: 'food',
    ....
  }, {... }]
}, {...}]
```
The way I'm looking at this, there could be repeated values (tags), but it would be for each entity (post), not in the same array...

J

Reply:
> The way I'm looking at this, there could be repeated values (tags), but it would be for each entity (post), not in the same array...

Ah, I think this is expected for this example because I assume `post_tag` has a unique `(post_id,tag_id)` pair, right? It cannot repeat here (unless the post has no tag returning `null`). But let's say we want to get the `users`  and the `tags` of the `posts` they publish. If they have two or more posts with the same tag, they will repeat.

```http
GET /users?=select=name,...posts(...post_tag(tags(*)))
```
```javascript
[
  {
    "name": "user_1",
    "tags": 
      [{"name": "food", ...}, 
       {"name": "food", ...},
       {"name": "tech", ...},
       {"name": "food", ...},
       null,
       null,
       {...}]
  }, 
  {...}
]
```

> I don't immediately understand this conclusion. Why are repeated values bad?

Hmm, yeah, this may be biased towards my personal preferences here. From the example above, I don't find it very useful to get the repeated tags or `null`s. But yes, I can see that repeating the values is the result as-is.

Reply:
In reality there should be options for `distinct` and `not null`, but I would think both of them should be true by default.

Reply:
`*,tags!post_tag(*)` would not return the same thing as `*,...post_tag(tags(*))`. As you know, `tags` is a completely different table than `post_tag`. I don't want the junction table, but the tag info.
> We could do something "magic" like .. vs .... The shorter one would return fewer values, so would turn both distinct and not null to ON. The longer one would turn both off. WDYT?

That could work, interesting concept.

J

Reply:
Yup, I see it now. You're coming from the other direction. I was thinking you were skipping `roles` which didn't make sense. 

I guess in this particular case, it would just be adding the ability to do it more than one way.